File: backend/src/routes/requests.js
const express = require("express");
const { ObjectId } = require("mongodb");

const { RequestCreate, RequestUpdate } = require("../validation/schemas");
const { validateBody } = require("../middleware/validate");
const { requestsCol, tenantsCol } = require("../database");
const { requireOwner, requireTenant, getCurrentUser } = require("../auth");
const { serialize, serializeList, toObjectId } = require("../utils/helpers");
const { asyncHandler } = require("../utils/asyncHandler");
const { HttpError } = require("../utils/httpError");

const router = express.Router();

router.post(
  "",
  requireTenant,
  validateBody(RequestCreate),
  asyncHandler(async (req, res) => {
    const payload = req.body;
    const tenant = await tenantsCol.findOne({ _id: new ObjectId(req.user.tenant_id) });
    const doc = {
      tenant_id: new ObjectId(req.user.tenant_id),
      room_id: tenant ? tenant.room_id || null : null,
      title: payload.title,
      description: payload.description,
      category: payload.category,
      status: "open",
      owner_note: null,
      created_at: new Date(),
      resolved_at: null,
    };
    const result = await requestsCol.insertOne(doc);
    const reqDoc = await requestsCol.findOne({ _id: result.insertedId });
    res.json(serialize(reqDoc));
  })
);

router.get(
  "",
  getCurrentUser,
  asyncHandler(async (req, res) => {
    const query = {};
    if (req.user.role === "tenant") {
      query.tenant_id = new ObjectId(req.user.tenant_id);
    }
    const reqs = await requestsCol.find(query).sort({ created_at: -1 }).limit(1000).toArray();

    const tenantIds = [...new Set(reqs.map((r) => String(r.tenant_id)))].map((id) => new ObjectId(id));
    const tenants = {};
    if (tenantIds.length) {
      const cursor = tenantsCol.find({ _id: { $in: tenantIds } });
      for await (const t of cursor) {
        tenants[String(t._id)] = t.name;
      }
    }
    for (const r of reqs) {
      r.tenant_name = tenants[String(r.tenant_id)];
    }

    res.json(serializeList(reqs));
  })
);

router.put(
  "/:requestId",
  requireOwner,
  validateBody(RequestUpdate),
  asyncHandler(async (req, res) => {
    const oid = toObjectId(req.params.requestId);
    const payload = req.body;
    const updates = { status: payload.status };
    if (payload.owner_note !== null && payload.owner_note !== undefined) {
      updates.owner_note = payload.owner_note;
    }
    if (payload.status === "resolved") {
      updates.resolved_at = new Date();
    }
    const result = await requestsCol.updateOne({ _id: oid }, { $set: updates });
    if (result.matchedCount === 0) {
      throw new HttpError(404, "Request not found");
    }
    const reqDoc = await requestsCol.findOne({ _id: oid });
    res.json(serialize(reqDoc));
  })
);

module.exports = router;

File: backend/app/routers/requests.py
from fastapi import APIRouter, Depends, HTTPException
from datetime import datetime
from bson import ObjectId

from app.models import RequestCreate, RequestUpdate
from app.database import requests_col, tenants_col
from app.auth import require_owner, require_tenant, get_current_user
from app.utils.helpers import serialize, serialize_list, to_object_id

router = APIRouter(prefix="/api/requests", tags=["requests"])


@router.post("")
async def create_request(payload: RequestCreate, user: dict = Depends(require_tenant)):
    tenant = await tenants_col.find_one({"_id": ObjectId(user["tenant_id"])})
    doc = {
        "tenant_id": ObjectId(user["tenant_id"]),
        "room_id": tenant.get("room_id") if tenant else None,
        "title": payload.title,
        "description": payload.description,
        "category": payload.category,
        "status": "open",
        "owner_note": None,
        "created_at": datetime.utcnow(),
        "resolved_at": None,
    }
    result = await requests_col.insert_one(doc)
    req = await requests_col.find_one({"_id": result.inserted_id})
    return serialize(req)


@router.get("")
async def list_requests(user: dict = Depends(get_current_user)):
    query = {}
    if user["role"] == "tenant":
        query["tenant_id"] = ObjectId(user["tenant_id"])
    reqs = await requests_col.find(query).sort("created_at", -1).to_list(1000)

    tenant_ids = {r["tenant_id"] for r in reqs}
    tenants = {}
    if tenant_ids:
        async for t in tenants_col.find({"_id": {"$in": list(tenant_ids)}}):
            tenants[t["_id"]] = t["name"]
    for r in reqs:
        r["tenant_name"] = tenants.get(r["tenant_id"])

    return serialize_list(reqs)


@router.put("/{request_id}")
async def update_request(request_id: str, payload: RequestUpdate, owner: dict = Depends(require_owner)):
    oid = to_object_id(request_id)
    updates = {"status": payload.status}
    if payload.owner_note is not None:
        updates["owner_note"] = payload.owner_note
    if payload.status == "resolved":
        updates["resolved_at"] = datetime.utcnow()
    result = await requests_col.update_one({"_id": oid}, {"$set": updates})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="Request not found")
    req = await requests_col.find_one({"_id": oid})
    return serialize(req)
